# Part B: Dataset Understanding & Preparation

## Task 6: Load the Dataset from CSV

In [17]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [7]:
import pandas as pd

df = pd.read_csv("../dataset/Advanced_Regression_HousePrice_Dataset_3800(Sheet1).csv")

print(df.head())

   property_id  sale_date  area_sqft  bedrooms  bathrooms  location_score  \
0       200001   1/1/2014       2181         6          4             8.1   
1       200002  12/1/2019       2383         5          4             5.3   
2       200003  10/1/2016       1047         3          3             5.9   
3       200004   3/1/2013       1753         4          3             7.0   
4       200005   7/1/2013       1728         4          4            10.0   

   property_age  distance_city_km  near_school  near_metro  crime_rate_index  \
0            21               3.8            0           0              4.84   
1            28              10.9            1           1              2.89   
2             7              27.5            0           1              4.04   
3            27              12.1            0           0              3.28   
4            32               1.4            0           1              3.84   

   house_price_inr  
0         35154898  
1         2671

In [8]:
# Check dataset shape
print("Dataset Shape:", df.shape)

# Check column names
print("\nColumn Names:")
print(df.columns.tolist())

# Check dataset information
df.info()

Dataset Shape: (3800, 12)

Column Names:
['property_id', 'sale_date', 'area_sqft', 'bedrooms', 'bathrooms', 'location_score', 'property_age', 'distance_city_km', 'near_school', 'near_metro', 'crime_rate_index', 'house_price_inr']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3800 entries, 0 to 3799
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   property_id       3800 non-null   int64  
 1   sale_date         3800 non-null   object 
 2   area_sqft         3800 non-null   int64  
 3   bedrooms          3800 non-null   int64  
 4   bathrooms         3800 non-null   int64  
 5   location_score    3800 non-null   float64
 6   property_age      3800 non-null   int64  
 7   distance_city_km  3800 non-null   float64
 8   near_school       3800 non-null   int64  
 9   near_metro        3800 non-null   int64  
 10  crime_rate_index  3800 non-null   float64
 11  house_price_inr   3800 non-null   int64  
dtype

### Interpretation

The house price dataset was successfully loaded from the CSV file using
the Pandas library. The dataset contains 3800 records and 12 columns.
The `head()` function was used to display the first five rows and
understand the structure of the dataset.

## Task 7 – Identify Features & Target Variable

In [10]:
# Define target variable
y = df["house_price_inr"]

# Define input features
X = df.drop(columns=["house_price_inr"])

# Display feature names
print("Input Features:")
print(X.columns.tolist())

# Display target variable
print("\nTarget Variable:")
print("house_price_inr")

# Check shapes
print("\nFeatures Shape:", X.shape)
print("Target Shape:", y.shape)


Input Features:
['property_id', 'sale_date', 'area_sqft', 'bedrooms', 'bathrooms', 'location_score', 'property_age', 'distance_city_km', 'near_school', 'near_metro', 'crime_rate_index']

Target Variable:
house_price_inr

Features Shape: (3800, 11)
Target Shape: (3800,)


## Task 8 – Missing Value Treatment & Basic Preprocessing

In [12]:

import numpy as np
# Check missing values
print("Missing values before treatment:\n", X.isnull().sum())

# Fill numeric missing values with median
for col in X.select_dtypes(include=[np.number]).columns:
    X[col] = X[col].fillna(X[col].median())

# Fill categorical/binary missing values with mode
for col in X.select_dtypes(exclude=[np.number]).columns:
    X[col] = X[col].fillna(X[col].mode()[0])

# Convert sale_date into datetime format
X["sale_date"] = pd.to_datetime(X["sale_date"])

# Extract useful features from sale_date
X["sale_year"] = X["sale_date"].dt.year
X["sale_month"] = X["sale_date"].dt.month

# Remove original sale_date column
X = X.drop(columns=["sale_date"])

# Remove property_id because it is only an identifier
X = X.drop(columns=["property_id"], errors="ignore")

# Check missing values after treatment
print("\nMissing values after treatment:\n", X.isnull().sum())

# Check data types
print("\nData Types:\n", X.dtypes)

# Scale continuous numeric features
from sklearn.preprocessing import StandardScaler

num_cols = [
    "area_sqft",
    "bedrooms",
    "bathrooms",
    "location_score",
    "property_age",
    "distance_city_km",
    "crime_rate_index",
    "sale_year",
    "sale_month"
]

scaler = StandardScaler()

X[num_cols] = scaler.fit_transform(X[num_cols])

# Show sample after preprocessing
X.head()

Missing values before treatment:
 property_id         0
sale_date           0
area_sqft           0
bedrooms            0
bathrooms           0
location_score      0
property_age        0
distance_city_km    0
near_school         0
near_metro          0
crime_rate_index    0
dtype: int64

Missing values after treatment:
 area_sqft           0
bedrooms            0
bathrooms           0
location_score      0
property_age        0
distance_city_km    0
near_school         0
near_metro          0
crime_rate_index    0
sale_year           0
sale_month          0
dtype: int64

Data Types:
 area_sqft             int64
bedrooms              int64
bathrooms             int64
location_score      float64
property_age          int64
distance_city_km    float64
near_school           int64
near_metro            int64
crime_rate_index    float64
sale_year             int32
sale_month            int32
dtype: object


,area_sqft,bedrooms,bathrooms,location_score,property_age,distance_city_km,near_school,near_metro,crime_rate_index,sale_year,sale_month
0,0.796121,1.895935,0.956143,0.904371,-0.124723,-1.420491,0,0,0.291961,-0.608266,-1.581503
1,1.142652,1.158745,0.956143,-0.680494,0.443269,-0.334294,1,1,-0.661537,0.631362,1.594267
2,-1.149258,-0.315633,0.073835,-0.340880,-1.260707,2.205267,0,1,-0.099218,-0.112415,1.016854
3,0.061886,0.421556,0.073835,0.281746,0.362127,-0.150711,0,0,-0.470837,-0.856191,-1.004090
4,0.018998,0.421556,0.956143,1.979815,0.767835,-1.787656,0,1,-0.197012,-0.856191,0.150735


In [13]:
# Prepare features and target variable

# Create a clean copy
df_clean = df.copy()

# Drop rows where target value is missing
df_clean = df_clean.dropna(subset=["house_price_inr"])

# Define features and target variable
X = df_clean.drop(columns=["house_price_inr"])
y = df_clean["house_price_inr"]

# Remove property_id if present
X = X.drop(columns=["property_id"], errors="ignore")

# Convert sale_date into useful numeric features
X["sale_date"] = pd.to_datetime(X["sale_date"])

X["sale_year"] = X["sale_date"].dt.year
X["sale_month"] = X["sale_date"].dt.month

# Remove original sale_date column
X = X.drop(columns=["sale_date"])

# Split the dataset into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Display dataset shapes
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (3040, 11)
X_test shape: (760, 11)
y_train shape: (3040,)
y_test shape: (760,)


### Interpretation

The dataset contains 11 input features and one target variable.
The target variable is `house_price_inr`, which represents the price
of the property. The remaining columns are used as input features
for predicting house prices.

# Part C: Regularized Linear Models

## Task 9 – Implement Ridge Regression (L2)

In [14]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error, r2_score

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Create Ridge Regression model
ridge_model = Ridge(alpha=1.0)

# Train the model
ridge_model.fit(X_train, y_train)

# Make predictions
y_pred_ridge = ridge_model.predict(X_test)

# Display first 5 predictions
print("First 5 Ridge Predictions:")
print(y_pred_ridge[:5])


First 5 Ridge Predictions:
[10797244.08389726 25651553.71833006 23085746.48837522 21709739.67983687
 30856954.39044386]


## Task 10 — Implement Lasso Regression (L1)

In [17]:
# Import Lasso Regression
from sklearn.linear_model import Lasso

# Create Lasso Regression model
lasso_model = Lasso(
    alpha=1.0,
    max_iter=10000
)

# Train the model
lasso_model.fit(X_train, y_train)

# Make predictions
y_pred_lasso = lasso_model.predict(X_test)

# Display first 5 predictions
print("First 5 Lasso Predictions:")
print(y_pred_lasso[:5])

First 5 Lasso Predictions:
[10796421.60363424 25651409.25936085 23085975.62823811 21709669.91669694
 30857683.43948025]


## Task 11 — Tune the Regularization Parameter (Alpha) Using Cross-Validation

In [18]:
# using Cross-Validation

from sklearn.model_selection import GridSearchCV

# Define different alpha values
param_grid = {
    "alpha": [0.01, 0.1, 1, 10, 100, 1000]
}

# Create Ridge model
ridge = Ridge()

# Perform Grid Search with 5-Fold Cross-Validation
grid_search = GridSearchCV(
    estimator=ridge,
    param_grid=param_grid,
    cv=5,
    scoring="neg_mean_squared_error"
)

# Train Grid Search
grid_search.fit(X_train, y_train)

# Display the best alpha value
print("Best Alpha:", grid_search.best_params_["alpha"])

# Display the best cross-validation score
print("Best Cross-Validation Score:", grid_search.best_score_)

Best Alpha: 10
Best Cross-Validation Score: -6221951335644.521


## Task 12 — Compare Ridge and Lasso Models

In [19]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# Ridge Regression evaluation
ridge_mae = mean_absolute_error(y_test, y_pred_ridge)
ridge_mse = mean_squared_error(y_test, y_pred_ridge)
ridge_rmse = np.sqrt(ridge_mse)
ridge_r2 = r2_score(y_test, y_pred_ridge)

# Lasso Regression evaluation
lasso_mae = mean_absolute_error(y_test, y_pred_lasso)
lasso_mse = mean_squared_error(y_test, y_pred_lasso)
lasso_rmse = np.sqrt(lasso_mse)
lasso_r2 = r2_score(y_test, y_pred_lasso)

# Create comparison table
comparison = pd.DataFrame({
    "Model": ["Ridge Regression", "Lasso Regression"],
    "MAE": [ridge_mae, lasso_mae],
    "MSE": [ridge_mse, lasso_mse],
    "RMSE": [ridge_rmse, lasso_rmse],
    "R2 Score": [ridge_r2, lasso_r2]
})

print(comparison)

              Model           MAE           MSE          RMSE  R2 Score
0  Ridge Regression  1.945453e+06  6.449981e+12  2.539681e+06  0.919911
1  Lasso Regression  1.945520e+06  6.450121e+12  2.539709e+06  0.919909


# Part D: Cross-Validation Strategies

## Task 13 — Apply and Compare Cross-Validation Techniques

In [21]:
# Task 13 — Apply and Compare Cross-Validation Techniques

import numpy as np
import pandas as pd

from sklearn.model_selection import (
    KFold,
    StratifiedKFold,
    LeaveOneOut,
    TimeSeriesSplit,
    cross_val_score
)

from sklearn.linear_model import Ridge
from sklearn.preprocessing import KBinsDiscretizer

# Create the model
model = Ridge(alpha=1.0)

# --------------------------------------------------
# 1. K-Fold Cross-Validation
# --------------------------------------------------

kfold = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

kfold_scores = cross_val_score(
    model,
    X,
    y,
    cv=kfold,
    scoring="r2"
)

print("K-Fold CV Scores:", kfold_scores)
print("K-Fold Mean R²:", kfold_scores.mean())


# --------------------------------------------------
# 2. Stratified K-Fold Cross-Validation
#    (Binning target values)
# --------------------------------------------------

# Convert continuous target into bins
y_binned = pd.qcut(
    y,
    q=5,
    labels=False,
    duplicates="drop"
)

stratified_kfold = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

stratified_scores = []

for train_index, test_index in stratified_kfold.split(X, y_binned):

    X_train_cv = X.iloc[train_index]
    X_test_cv = X.iloc[test_index]

    y_train_cv = y.iloc[train_index]
    y_test_cv = y.iloc[test_index]

    model.fit(X_train_cv, y_train_cv)

    score = model.score(X_test_cv, y_test_cv)

    stratified_scores.append(score)

print("\nStratified K-Fold CV Scores:", stratified_scores)
print("Stratified K-Fold Mean R²:", np.mean(stratified_scores))

# --------------------------------------------------
# 3. Leave-One-Out Cross-Validation
# --------------------------------------------------

from sklearn.metrics import mean_squared_error

loo = LeaveOneOut()

# Use negative mean squared error because
# R² cannot be calculated with only one test sample
loo_scores = cross_val_score(
    model,
    X,
    y,
    cv=loo,
    scoring="neg_mean_squared_error"
)

# Convert negative MSE to positive MSE
loo_mse = -loo_scores.mean()

# Calculate RMSE
loo_rmse = np.sqrt(loo_mse)

print("\nLeave-One-Out CV Mean MSE:", loo_mse)
print("Leave-One-Out CV RMSE:", loo_rmse)


# --------------------------------------------------
# 4. Time Series Split
# --------------------------------------------------

# Sort data using a time-related feature
# Here sale_year is used as the time-related feature

if "sale_year" in X.columns:

    sorted_index = X["sale_year"].sort_values().index

    X_time = X.loc[sorted_index]
    y_time = y.loc[sorted_index]

    time_series_cv = TimeSeriesSplit(n_splits=5)

    time_scores = cross_val_score(
        model,
        X_time,
        y_time,
        cv=time_series_cv,
        scoring="r2"
    )

    print("\nTime Series Split Scores:", time_scores)
    print("Time Series Split Mean R²:", time_scores.mean())

else:
    print("\nTime Series Split could not run because sale_year is not available.")

K-Fold CV Scores: [0.919911   0.91604889 0.91281652 0.92028253 0.91817738]
K-Fold Mean R²: 0.9174472649502448

Stratified K-Fold CV Scores: [0.9196347269502454, 0.9203904866867364, 0.9213882597312097, 0.9136620515296273, 0.9124879810627822]
Stratified K-Fold Mean R²: 0.9175127011921201

Leave-One-Out CV Mean MSE: 6255435523988.049
Leave-One-Out CV RMSE: 2501086.8685409646

Time Series Split Scores: [0.91279176 0.91697085 0.92486848 0.90881585 0.92173747]
Time Series Split Mean R²: 0.9170368842131099


### Interpretation

K-Fold Cross-Validation achieved a mean R² score of approximately 0.9174.
Stratified K-Fold achieved a mean R² score of approximately 0.9175, which
was slightly higher than the K-Fold result. Time Series Split achieved a
mean R² score of approximately 0.9170.

Leave-One-Out Cross-Validation uses only one sample for testing in each
iteration. Therefore, R² cannot be calculated because it requires at least
two test samples. For this reason, MSE and RMSE were used to evaluate the
Leave-One-Out Cross-Validation performance.

## Task 14 — Analyze Performance Metrics Across Different CV Strategies

In [22]:
cv_results = pd.DataFrame({
    
    "CV Strategy": [
        "K-Fold",
        "Stratified K-Fold",
        "Time Series Split"
    ],
    
    "Mean R² Score": [
        kfold_scores.mean(),
        np.mean(stratified_scores),
        time_scores.mean()
    ],
    
    "Standard Deviation": [
        kfold_scores.std(),
        np.std(stratified_scores),
        time_scores.std()
    ]
})

print(cv_results)

# Display Leave-One-Out result separately
print("\nLeave-One-Out Cross-Validation Results")
print("Mean MSE:", loo_mse)
print("RMSE:", loo_rmse)

         CV Strategy  Mean R² Score  Standard Deviation
0             K-Fold       0.917447            0.002758
1  Stratified K-Fold       0.917513            0.003685
2  Time Series Split       0.917037            0.005814

Leave-One-Out Cross-Validation Results
Mean MSE: 6255435523988.049
RMSE: 2501086.8685409646


# Part E: Tree-Based Regression Models

## Task 15 — Implement Decision Tree Regression

In [23]:
from sklearn.tree import DecisionTreeRegressor

# Create Decision Tree Regression model
tree_model = DecisionTreeRegressor(
    random_state=42
)

# Train the model
tree_model.fit(X_train, y_train)

# Make predictions
y_pred_tree = tree_model.predict(X_test)

# Display first 5 predictions
print("First 5 Decision Tree Predictions:")
print(y_pred_tree[:5])

First 5 Decision Tree Predictions:
[12257054. 19803272. 23447175. 19673601. 33610120.]


## Task 16 — Control Tree Complexity

In [24]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Create a controlled Decision Tree model
tree_controlled = DecisionTreeRegressor(
    max_depth=10,
    min_samples_split=10,
    min_samples_leaf=5,
    random_state=42
)

# Train the model
tree_controlled.fit(X_train, y_train)

# Make predictions
y_pred_tree_controlled = tree_controlled.predict(X_test)

# Calculate performance metrics
tree_mae = mean_absolute_error(
    y_test,
    y_pred_tree_controlled
)

tree_mse = mean_squared_error(
    y_test,
    y_pred_tree_controlled
)

tree_rmse = np.sqrt(tree_mse)

tree_r2 = r2_score(
    y_test,
    y_pred_tree_controlled
)

# Display results
print("Controlled Decision Tree Results")
print("MAE:", tree_mae)
print("MSE:", tree_mse)
print("RMSE:", tree_rmse)
print("R² Score:", tree_r2)

Controlled Decision Tree Results
MAE: 2155390.8359593656
MSE: 8725474277219.505
RMSE: 2953891.3787103794
R² Score: 0.8916563490057925


## Task 17 — Implement Random Forest Regression

In [25]:
from sklearn.ensemble import RandomForestRegressor

# Create Random Forest model
forest_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

# Train the model
forest_model.fit(X_train, y_train)

# Make predictions
y_pred_forest = forest_model.predict(X_test)

# Calculate performance metrics
forest_mae = mean_absolute_error(
    y_test,
    y_pred_forest
)

forest_mse = mean_squared_error(
    y_test,
    y_pred_forest
)

forest_rmse = np.sqrt(forest_mse)

forest_r2 = r2_score(
    y_test,
    y_pred_forest
)

# Display results
print("Random Forest Results")
print("MAE:", forest_mae)
print("MSE:", forest_mse)
print("RMSE:", forest_rmse)
print("R² Score:", forest_r2)

Random Forest Results
MAE: 1757755.0067894738
MSE: 5785958403336.219
RMSE: 2405401.9213711913
R² Score: 0.9281561279076027


## Task 18 — Compare Decision Tree and Random Forest

In [26]:
comparison_tree_models = pd.DataFrame({

    "Model": [
        "Decision Tree",
        "Random Forest"
    ],

    "MAE": [
        tree_mae,
        forest_mae
    ],

    "MSE": [
        tree_mse,
        forest_mse
    ],

    "RMSE": [
        tree_rmse,
        forest_rmse
    ],

    "R² Score": [
        tree_r2,
        forest_r2
    ]

})

print(comparison_tree_models)

           Model           MAE           MSE          RMSE  R² Score
0  Decision Tree  2.155391e+06  8.725474e+12  2.953891e+06  0.891656
1  Random Forest  1.757755e+06  5.785958e+12  2.405402e+06  0.928156


# Part F: Support Vector Regression

## Task 19 — Implement Support Vector Regression

In [27]:
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler

# Create scaler
scaler = StandardScaler()

# Scale training and testing features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# -------------------------------
# Linear Kernel SVR
# -------------------------------

svr_linear = SVR(
    kernel="linear",
    C=100
)

# Train Linear SVR
svr_linear.fit(
    X_train_scaled,
    y_train
)

# Make predictions
y_pred_svr_linear = svr_linear.predict(
    X_test_scaled
)

# -------------------------------
# RBF Kernel SVR
# -------------------------------

svr_rbf = SVR(
    kernel="rbf",
    C=100,
    epsilon=0.1
)

# Train RBF SVR
svr_rbf.fit(
    X_train_scaled,
    y_train
)

# Make predictions
y_pred_svr_rbf = svr_rbf.predict(
    X_test_scaled
)

print("First 5 Linear SVR Predictions:")
print(y_pred_svr_linear[:5])

print("\nFirst 5 RBF SVR Predictions:")
print(y_pred_svr_rbf[:5])

First 5 Linear SVR Predictions:
[19489377.20279558 20870801.51300517 20179451.59847296 20002591.05109671
 20779342.32332663]

First 5 RBF SVR Predictions:
[20042618.76476206 20065103.32003241 20057837.22303831 20052556.44065307
 20068625.91355181]


## Task 20 — Tune SVR Hyperparameters

In [28]:
from sklearn.model_selection import GridSearchCV

# Define SVR model
svr_model = SVR()

# Define hyperparameters
param_grid = {

    "kernel": ["linear", "rbf"],

    "C": [
        1,
        10,
        100
    ],

    "gamma": [
        "scale",
        "auto"
    ],

    "epsilon": [
        0.01,
        0.1,
        1
    ]
}

# Create Grid Search with 5-Fold Cross-Validation
grid_search_svr = GridSearchCV(
    estimator=svr_model,
    param_grid=param_grid,
    cv=5,
    scoring="r2",
    n_jobs=-1
)

# Train Grid Search
grid_search_svr.fit(
    X_train_scaled,
    y_train
)

# Display best parameters
print("Best Parameters:")
print(grid_search_svr.best_params_)

# Display best score
print("\nBest Cross-Validation R² Score:")
print(grid_search_svr.best_score_)

Best Parameters:
{'C': 100, 'epsilon': 0.1, 'gamma': 'scale', 'kernel': 'linear'}

Best Cross-Validation R² Score:
0.0758738929722687


# Task 21 — Compare SVR with Linear and Tree-Based Models

In [29]:
# Calculate SVR Linear performance
svr_linear_mae = mean_absolute_error(
    y_test,
    y_pred_svr_linear
)

svr_linear_rmse = np.sqrt(
    mean_squared_error(
        y_test,
        y_pred_svr_linear
    )
)

svr_linear_r2 = r2_score(
    y_test,
    y_pred_svr_linear
)

# Calculate SVR RBF performance
svr_rbf_mae = mean_absolute_error(
    y_test,
    y_pred_svr_rbf
)

svr_rbf_rmse = np.sqrt(
    mean_squared_error(
        y_test,
        y_pred_svr_rbf
    )
)

svr_rbf_r2 = r2_score(
    y_test,
    y_pred_svr_rbf
)

# Create final comparison table
final_comparison = pd.DataFrame({

    "Model": [
        "Ridge Regression",
        "Decision Tree",
        "Random Forest",
        "SVR Linear",
        "SVR RBF"
    ],

    "R² Score": [
        ridge_r2,
        tree_r2,
        forest_r2,
        svr_linear_r2,
        svr_rbf_r2
    ],

    "MAE": [
        ridge_mae,
        tree_mae,
        forest_mae,
        svr_linear_mae,
        svr_rbf_mae
    ],

    "RMSE": [
        ridge_rmse,
        tree_rmse,
        forest_rmse,
        svr_linear_rmse,
        svr_rbf_rmse
    ]

})

print(final_comparison)

              Model  R² Score           MAE          RMSE
0  Ridge Regression  0.919911  1.945453e+06  2.539681e+06
1     Decision Tree  0.891656  2.155391e+06  2.953891e+06
2     Random Forest  0.928156  1.757755e+06  2.405402e+06
3        SVR Linear  0.098344  6.620502e+06  8.521443e+06
4           SVR RBF -0.000627  6.985495e+06  8.976954e+06


### Interpretation

Different regression models were compared using MAE, RMSE, and R² Score.
The model with the lowest MAE and RMSE and the highest R² Score can be
considered the best-performing model.

Linear models provide simple and interpretable predictions, while tree-based
ensemble models such as Random Forest can capture more complex relationships.
SVR performance depends on the selected kernel and hyperparameter values.

# Part G: Model Comparison & Evaluation

## Task 22 — Evaluate All Models Using Regression Metrics

In [30]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import pandas as pd
import numpy as np

# Create a dictionary containing predictions
models_predictions = {
    "Ridge Regression": y_pred_ridge,
    "Lasso Regression": y_pred_lasso,
    "Decision Tree": y_pred_tree_controlled,
    "Random Forest": y_pred_forest,
    "SVR Linear": y_pred_svr_linear,
    "SVR RBF": y_pred_svr_rbf
}

# Store evaluation results
results = []

# Calculate metrics for each model
for model_name, predictions in models_predictions.items():

    mse = mean_squared_error(y_test, predictions)

    mae = mean_absolute_error(y_test, predictions)

    rmse = np.sqrt(mse)

    r2 = r2_score(y_test, predictions)

    results.append([
        model_name,
        mse,
        mae,
        rmse,
        r2
    ])

# Create evaluation table
evaluation_table = pd.DataFrame(
    results,
    columns=[
        "Model",
        "MSE",
        "MAE",
        "RMSE",
        "R² Score"
    ]
)

# Display results
print(evaluation_table)

              Model           MSE           MAE          RMSE  R² Score
0  Ridge Regression  6.449981e+12  1.945453e+06  2.539681e+06  0.919911
1  Lasso Regression  6.450121e+12  1.945520e+06  2.539709e+06  0.919909
2     Decision Tree  8.725474e+12  2.155391e+06  2.953891e+06  0.891656
3     Random Forest  5.785958e+12  1.757755e+06  2.405402e+06  0.928156
4        SVR Linear  7.261499e+13  6.620502e+06  8.521443e+06  0.098344
5           SVR RBF  8.058569e+13  6.985495e+06  8.976954e+06 -0.000627


# Task 23 — Compare Different Model Categories

In [31]:
# Add model category
evaluation_table["Model Category"] = [
    "Regularized Linear Model",
    "Regularized Linear Model",
    "Tree-Based Model",
    "Tree-Based Ensemble",
    "Support Vector Regression",
    "Support Vector Regression"
]

# Rearrange columns
comparison_table = evaluation_table[[
    "Model Category",
    "Model",
    "MSE",
    "MAE",
    "RMSE",
    "R² Score"
]]

# Sort models by R² Score
comparison_table = comparison_table.sort_values(
    by="R² Score",
    ascending=False
)

print(comparison_table)

              Model Category             Model           MSE           MAE  \
3        Tree-Based Ensemble     Random Forest  5.785958e+12  1.757755e+06   
0   Regularized Linear Model  Ridge Regression  6.449981e+12  1.945453e+06   
1   Regularized Linear Model  Lasso Regression  6.450121e+12  1.945520e+06   
2           Tree-Based Model     Decision Tree  8.725474e+12  2.155391e+06   
4  Support Vector Regression        SVR Linear  7.261499e+13  6.620502e+06   
5  Support Vector Regression           SVR RBF  8.058569e+13  6.985495e+06   

           RMSE  R² Score  
3  2.405402e+06  0.928156  
0  2.539681e+06  0.919911  
1  2.539709e+06  0.919909  
2  2.953891e+06  0.891656  
4  8.521443e+06  0.098344  
5  8.976954e+06 -0.000627  


# Task 24 — Check Overfitting and Underfitting

In [33]:
# Ridge Regression
ridge_train_r2 = ridge_model.score(X_train, y_train)
ridge_test_r2 = ridge_model.score(X_test, y_test)

# Lasso Regression
lasso_train_r2 = lasso_model.score(X_train, y_train)
lasso_test_r2 = lasso_model.score(X_test, y_test)

# Decision Tree
tree_train_r2 = tree_controlled.score(X_train, y_train)
tree_test_r2 = tree_controlled.score(X_test, y_test)

# Random Forest
forest_train_r2 = forest_model.score(X_train, y_train)
forest_test_r2 = forest_model.score(X_test, y_test)

# Create comparison table
overfitting_table = pd.DataFrame({
    "Model": [
        "Ridge Regression",
        "Lasso Regression",
        "Decision Tree",
        "Random Forest"
    ],
    "Training R²": [
        ridge_train_r2,
        lasso_train_r2,
        tree_train_r2,
        forest_train_r2
    ],
    "Testing R²": [
        ridge_test_r2,
        lasso_test_r2,
        tree_test_r2,
        forest_test_r2
    ]
})

# Calculate difference
overfitting_table["Difference"] = (
    overfitting_table["Training R²"]
    - overfitting_table["Testing R²"]
)

print(overfitting_table)

              Model  Training R²  Testing R²  Difference
0  Ridge Regression     0.917402    0.919911   -0.002509
1  Lasso Regression     0.917402    0.919909   -0.002507
2     Decision Tree     0.962133    0.891656    0.070476
3     Random Forest     0.989346    0.928156    0.061190


### Interpretation

The Decision Tree achieved a training R² score of 0.9621 and a testing R²
score of 0.8917. The difference of approximately 0.0705 indicates some
overfitting.

The Random Forest achieved a training R² score of 0.9893 and a testing R²
score of 0.9282. Although there is a small difference between training and
testing performance, the model generalizes better than the single Decision
Tree.

Therefore, Random Forest provides better predictive performance and
generalization among the tree-based models.

## Part H: Final Analysis & Reporting

## Task 25: Final Report

### 1. Best-Performing Model

The best-performing model was identified by comparing MSE, MAE, RMSE,
and R² Score across all regression models. The model with the lowest error
values and highest R² Score was considered the best model for house price
prediction.

### 2. Impact of Regularization

Ridge and Lasso Regression were used to control model complexity.
Ridge Regression uses L2 regularization and reduces the magnitude of
coefficients. Lasso Regression uses L1 regularization and can reduce some
coefficients to zero, which can perform automatic feature selection.

### 3. Role of Cross-Validation

Cross-validation was used to evaluate the stability and reliability of
the models. K-Fold, Stratified K-Fold, Leave-One-Out, and Time Series Split
were applied. The results showed that model performance remained relatively
stable across different validation strategies.

### 4. Comparison of Linear and Non-Linear Models

Regularized linear models provide simple and interpretable predictions.
Tree-based models and Support Vector Regression can capture more complex
and non-linear relationships. Random Forest combines multiple decision
trees and generally provides more stable predictions.

### 5. Business Interpretation

The regression models can help a real estate company estimate property
prices based on available property features. Accurate price predictions can
support pricing decisions, property valuation, investment analysis, and
market planning.